citest/gcp_testing/gcp_api.py

```diff
@@ -52,26 +55,84 @@ def build_authenticated_service(name, version, scopes, credentials_path):
   return discovery.build(name, version, http=http)
 
 
-class GoogleAgentObservationFailureVerifier(jc.ObservationFailureVerifier):
+class GoogleAgentObservationFailureVerifier(jc.ObservationVerifier):
   """An ObservationVerifier that expects specific errors from stderr."""
 
-  def __init__(self, title, error_regex):
-    """Constructs the clause with the acceptable error regex.
+  def __init__(self, title, http_code=None, error_regex=None):
+    """Constructs the clause with the acceptable status and/or error regex.
 
     Args:
       title: Verifier name for reporting purposes only.
-      error_regex: Regex pattern for errors we're looking for.
+      http_code: [int] The specific HTTP status code that is expected or None.
+      error_regex: [string] Regex pattern expected in error responses or None.
     """
     super(GoogleAgentObservationFailureVerifier, self).__init__(title)
+    self.__http_code = http_code
     self.__error_regex = error_regex
 
   def export_to_json_snapshot(self, snapshot, entity):
     """Implements JsonSnapshotable interface."""
-    snapshot.edge_builder.make_control(entity, 'Regex', self.__error_regex)
+    if self.__http_code:
+      snapshot.edge_builder.make_control(entity, 'HTTP Code', self.__http_code)
+    if self.__error_regex:
+      snapshot.edge_builder.make_control(entity, 'Regex', self.__error_regex)
     super(GoogleAgentObservationFailureVerifier, self).export_to_json_snapshot(
         snapshot, entity)
 
-  def _error_comment_or_none(self, error):
-    if (isinstance(error, HttpError)
-        and error.match_regex(self.__error_regex)):
-      return 'Error matches {0}'.format(self.__error_regex)
+  def __call__(self, observation):
+    """Check if observation contains the expected HTTP error.
+
+    Args:
+      observation: [Observation] The observation
+    """
+    message = None
+    error = None
+    for error in observation.errors:
+      if not isinstance(error, HttpError):
+        continue
+
+      # Build up the error confirming the expected error
+      message = ['Observed HttpError']
+      sep = 'with'  # Separator between confirmation phrases.
+      if self.__http_code:
+        if self.__http_code == error.resp.status:
+          message.append('{0} HTTP Code={1}'.format(sep, self.__http_code))
+          sep = 'and'
+        else:
+          # Not the expected error
+          message = None
+          continue
+
+      if self.__error_regex:
+        if re.search(self.__error_regex, error.content):
+          message.append('{0} error matches {1}'.format(
+              sep, self.__error_regex))
+        else:
+          # Not the expected error
+          message = None
+          continue
+
+      # Error matched.
+      break
+
+    # Error does not match.
+    if message:
+      error_result = jc.ObservationFailedError(valid=True, failures=[error])
```

Comment:
from L89, `error` will always be `None`. I'm not sure what the intended behavior is here.